pkg/sql/colexec/auto_incr.go

```diff
@@ -0,0 +1,307 @@
+// Copyright 2022 Matrix Origin
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package colexec
+
+import (
+	"context"
+
+	"github.com/matrixorigin/matrixone/pkg/container/batch"
+	"github.com/matrixorigin/matrixone/pkg/container/nulls"
+	"github.com/matrixorigin/matrixone/pkg/container/types"
+	"github.com/matrixorigin/matrixone/pkg/container/vector"
+	"github.com/matrixorigin/matrixone/pkg/pb/plan"
+	"github.com/matrixorigin/matrixone/pkg/sql/errors"
+	"github.com/matrixorigin/matrixone/pkg/vm/engine"
+	"github.com/matrixorigin/matrixone/pkg/vm/engine/tae/moengine"
+	"github.com/matrixorigin/matrixone/pkg/vm/process"
+)
+
+func UpdateInsertBatch(e engine.Engine, ctx context.Context, proc *process.Process, ColDefs []*plan.ColDef, bat *batch.Batch, namePre string) error {
+	incrNum, err := GetRangeFromAutoIncrTable(e, ctx, proc, bat, ColDefs, namePre)
+	if err != nil {
+		return err
+	}
+
+	if err = UpdateBatchImpl(ColDefs, bat, incrNum); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func UpdateInsertValueBatch(e engine.Engine, ctx context.Context, proc *process.Process, p *plan.InsertValues, bat *batch.Batch) error {
+	namePre := p.DbName + "_" + p.TblName + "_"
+	ColDefs := p.ExplicitCols
+	OrderColDefs(p.OrderAttrs, ColDefs)
+	return UpdateInsertBatch(e, ctx, proc, ColDefs, bat, namePre)
+}
+
+func GetRangeFromAutoIncrTable(e engine.Engine, ctx context.Context, proc *process.Process, bat *batch.Batch, ColDefs []*plan.ColDef, namePre string) ([]int64, error) {
+	taeEngine, ok := e.(moengine.TxnEngine)
```

Comment:
Why do you care the engine ...

Reply:
because I need to start a transaction, so I need the engine

Reply:
The moengine.TxnEngine will be deprecated. You must use a TxnClient to start a new transaction.
I suggest adding a method to the engine.Table interface to get what you want instead of starting a transaction.